crates/tako/src/server/core.rs

```diff
@@ -237,47 +250,64 @@ impl Core {
     }
 
     #[inline]
-    pub fn add_ready_to_assign(&mut self, task_ref: TaskRef) {
-        self.ready_to_assign.push(task_ref);
+    pub fn add_ready_to_assign(&mut self, task_id: TaskId) {
+        self.ready_to_assign.push(task_id);
     }
 
+    // TODO: move to TaskMap
     /// Removes a single task.
     /// It can still remain in [`ready_to_assign`], where it will remain until the scheduler picks
     /// it up.
     #[must_use]
-    pub fn remove_task(&mut self, task: &mut Task) -> TaskRuntimeState {
-        trace_task_remove(task.id);
-        assert!(!task.has_consumers());
-        assert!(self.tasks.remove(&task.id).is_some());
-        std::mem::replace(&mut task.state, TaskRuntimeState::Released)
+    pub fn remove_task(&mut self, task_id: TaskId) -> TaskRuntimeState {
+        trace_task_remove(task_id);
+
+        let state = {
+            let task = self.tasks.get_task_mut(task_id);
+            assert!(!task.has_consumers());
+            std::mem::replace(&mut task.state, TaskRuntimeState::Released)
```

Comment:
Maybe TaskRuntimeState::Released can be completely removed now, as it was mainly a marker that "you have a reference on a task that is removed"

Old state can be safely taken from removed without replacing

Reply:
Good idea.